File: src/engine/round.ts
/**
 * Проигрывание одного раунда: от пары сидов до итоговой выплаты.
 *
 * Порт `slotmath/round.py`, который остаётся эталоном. Соответствие
 * доказывается не глазами, а тестом `round.test.ts`: он прогоняет 26
 * раундов из `tests/fixtures/rounds.json`, сгенерированных Python, и
 * сверяет каждую сетку и каждую выплату.
 *
 * ПОРЯДОК ОБРАЩЕНИЙ К RNG ЗАФИКСИРОВАН И МЕНЯТЬ ЕГО НЕЛЬЗЯ:
 *
 *   1. Базовый спин — 5 обращений nextInt(baseReels[i].length), i = 0..4.
 *   2. Каждый фриспин — 5 обращений nextInt(freeReels[i].length), i = 0..4.
 *      Ретриггер увеличивает счётчик оставшихся спинов, но не порядок.
 *
 * Любая вставка лишнего обращения (например, «случайного» украшения в
 * анимации) сдвинет весь поток, и раунды перестанут воспроизводиться
 * офлайн-верификатором. Всё, что нужно клиенту для показа, выводится
 * из уже полученных стопов.
 */

import type { GameConfig } from "./config.js";
import { NUM_REELS, NUM_ROWS, PAYLINES } from "./paylines.js";
import { RoundRandom, hashServerSeed } from "./rng.js";

/**
 * Предохранитель от бесконечной серии фриспинов.
 *
 * При включённом ретриггере серия теоретически неограниченна. Вероятность
 * дойти до 200 спинов исчезающе мала, но «исчезающе мало» на потоке в
 * миллионы раундов случается, а зависший запрос — это заблокированный
 * баланс игрока. Значение совпадает с MAX_FREE_SPINS в round.py.
 */
export const MAX_FREE_SPINS = 200;

export interface WinDetail {
  /** Номер линии от 1. Отсутствует у выплаты за scatter. */
  line?: number;
  symbol: string;
  count: number;
  /** Пары [барабан, ряд] выигрышной серии. У scatter отсутствуют. */
  positions?: number[][];
  /** Выплата в ставках на линию (для scatter — в ставках на линию от общей ставки). */
  pay: number;
  scatter?: boolean;
}

export interface SpinRecord {
  index: number;
  free: boolean;
  reelStops: number[];
  grid: string[][];
  win: number;
  multiplier: number;
  scatterCount: number;
  triggeredFreeSpins: number;
  winDetails: WinDetail[];
}

export interface RoundRecord {
  serverSeedHash: string;
  clientSeed: string;
  nonce: number;
  betPerLine: number;
  lines: number;
  totalBet: number;
  totalWin: number;
  capped: boolean;
  drawCount: number;
  /** Тир бонус-игры, выбранный игроком до спина (1 — обычная серия). */
  bonusTier: number;
  /** Фактический множитель серии: делитель награды, не больше тира. */
  bonusMultiplier: number;
  spins: SpinRecord[];
}

/** Окно 5x3 в виде grid[reel][row]. Лента кольцевая. */
export function windowFromStops(
  reels: readonly (readonly string[])[],
  stops: readonly number[],
): string[][] {
  const window: string[][] = [];
  for (let reel = 0; reel < NUM_REELS; reel++) {
    const strip = reels[reel];
    const length = strip.length;
    const stop = stops[reel];
    const column: string[] = [];
    for (let row = 0; row < NUM_ROWS; row++) {
      column.push(strip[(stop + row) % length]);
    }
    window.push(column);
  }
  return window;
}

export interface LineEvaluation {
  total: number;
  details: WinDetail[];
}

/**
 * Выплата по всем линиям в ставках на линию плюс детализация.
 *
 * Правило классическое: серия считается слева направо и обязана начинаться
 * с первого барабана, wild замещает любой оплачиваемый символ, а по линии
 * берётся МАКСИМУМ по всем символам.
 *
 * Максимум здесь принципиален. Если брать первый подходящий символ, линия
 * вида wild-wild-wild-TEN-TEN оплатится как пять TEN (113) вместо трёх
 * CROWN (150). Ошибка выглядит безобидной, но на дистанции сдвигает RTP
 * на десятки процентных пунктов и ловится только сверкой с аналитикой.
 */
export function evaluateLines(cfg: GameConfig, grid: string[][]): LineEvaluation {
  let total = 0;
  const details: WinDetail[] = [];

  for (let lineIndex = 0; lineIndex < PAYLINES.length; lineIndex++) {
    const line = PAYLINES[lineIndex];

    const symbols: string[] = [];
    for (let reel = 0; reel < NUM_REELS; reel++) {
      symbols.push(grid[reel][line[reel]]);
    }

    let bestPay = 0;
    let bestSymbol: string | null = null;
    let bestRun = 0;

    for (const symbol of Object.keys(cfg.paytable)) {
      let run = 0;
      for (let reel = 0; reel < NUM_REELS; reel++) {
        if (symbols[reel] === symbol || symbols[reel] === cfg.wild) run++;
        else break;
      }
      const pay = run >= 3 ? cfg.paytable[symbol][String(run)] ?? 0 : 0;
      if (pay > bestPay) {
        bestPay = pay;
        bestSymbol = symbol;
        bestRun = run;
      }
    }

    if (bestPay > 0 && bestSymbol !== null) {
      total += bestPay;
      const positions: number[][] = [];
      for (let reel = 0; reel < bestRun; reel++) positions.push([reel, line[reel]]);
      details.push({
        line: lineIndex + 1,
        symbol: bestSymbol,
        count: bestRun,
        positions,
        pay: bestPay,
      });
    }
  }

  return { total, details };
}

/** Scatter платит по всему окну, позиция не важна. */
export function countScatters(cfg: GameConfig, grid: string[][]): number {
  let count = 0;
  for (let reel = 0; reel < NUM_REELS; reel++) {
    for (let row = 0; row < NUM_ROWS; row++) {
      if (grid[reel][row] === cfg.scatter) count++;
    }
  }
  return count;
}

/**
 * Тиры бонус-игры «Сундуки короны» (T-195).
 *
 * Игрок выбирает форму серии фриспинов ДО спина: те же деньги, но иначе
 * распределённые. Множитель обязан делить количество фриспинов нацело —
 * тогда произведение «спины × множитель» постоянно, и матожидание не
 * меняется, меняется только волатильность.
 *
 * Тиры намеренно живут в коде, а не в `config/game.json`: конфиг лент —
 * это математика, по которой считался PAR sheet, и его хэш участвует в
 * проверке раундов. Добавить туда поле значит инвалидировать все прошлые
 * проверки ради вещи, которая к лентам отношения не имеет.
 */
export const BONUS_TIERS = [1, 5, 25] as const;
export type BonusTier = (typeof BONUS_TIERS)[number];

/**
 * Наибольший делитель `award`, не превосходящий выбранный тир.
 *
 * Награда бывает 10, 15 или 25 фриспинов, и не всякий тир делит её нацело.
 * Округление вниз до делителя сохраняет равенство «спины × множитель =
 * награда», то есть матожидание. Игрок, выбравший «экстрим» при награде 15,
 * получит один спин с множителем 15, а не 25 — и это честно.
 */
export function effectiveBonusMultiplier(award: number, tier: number): number {
  if (award <= 0 || tier <= 1) return 1;
  let best = 1;
  for (let m = 1; m <= Math.min(tier, award); m += 1) {
    if (award % m === 0) best = m;
  }
  return best;
}

export interface PlayRoundOptions {
  betPerLine?: number;
  /** Выбранный игроком тир бонуса. По умолчанию 1 — обычная серия. */
  bonusTier?: number;
}

/**
 * Полный раунд: базовый спин плюс вся серия фриспинов.
 *
 * Возвращается детерминированно: те же сиды и nonce всегда дают тот же
 * результат. Именно на этом держится provably fair — игрок повторит
 * расчёт офлайн и получит то же самое.
 */
export function playRound(
  cfg: GameConfig,
  serverSeed: string,
  clientSeed: string,
  nonce: number,
  options: PlayRoundOptions = {},
): RoundRecord {
  const betPerLine = options.betPerLine ?? 1;
  if (!Number.isInteger(betPerLine) || betPerLine < 1) {
    throw new Error(`betPerLine должен быть целым от 1, получено ${betPerLine}`);
  }
  const bonusTier = options.bonusTier ?? 1;
  if (!BONUS_TIERS.includes(bonusTier as BonusTier)) {
    throw new Error(`bonusTier=${bonusTier} не входит в ${BONUS_TIERS.join(", ")}`);
  }

  const rng = new RoundRandom(serverSeed, clientSeed, nonce);
  const totalBet = betPerLine * cfg.lines;
  const spins: SpinRecord[] = [];

  // --- базовый спин ---
  const baseStops: number[] = [];
  for (let reel = 0; reel < NUM_REELS; reel++) {
    baseStops.push(rng.nextInt(cfg.baseReels[reel].length));
  }
  const baseGrid = windowFromStops(cfg.baseReels, baseStops);
  const baseEval = evaluateLines(cfg, baseGrid);
  const baseScatters = countScatters(cfg, baseGrid);

  // Выплата за scatter считается от ОБЩЕЙ ставки, а не от ставки на линию:
  // scatter не привязан к линии, поэтому и множитель применяется к целому.
  const baseScatterMult = cfg.scatterPays[String(baseScatters)] ?? 0;
  const baseScatterPay = baseScatterMult * totalBet;

  const baseDetails: WinDetail[] = [...baseEval.details];
  if (baseScatterPay > 0) {
    baseDetails.push({
      symbol: cfg.scatter,
      count: baseScatters,
      pay: baseScatterMult * cfg.lines,
      scatter: true,
    });
  }

  let triggered = 0;
  let remaining = 0;
  // Множитель бонуса: сжимает серию фриспинов, во столько же раз увеличивая
  // каждую выплату. При тире 1 (по умолчанию) множитель равен единице и
  // раунд считается ровно так же, как считался всегда, — иначе разошлись бы
  // проверки прошлых раундов и сверка с эталонной реализацией на Python.
  const bonusMultiplier = effectiveBonusMultiplier(
    cfg.freeSpinsAward[String(baseScatters)] ?? 0,
    bonusTier,
  );
  if (baseScatters >= cfg.scatterTrigger) {
    triggered = cfg.freeSpinsAward[String(baseScatters)] ?? 0;
    remaining = triggered / bonusMultiplier;
  }

  const baseWin = baseEval.total * betPerLine + baseScatterPay;
  spins.push({
    index: 0,
    free: false,
    reelStops: baseStops,
    grid: baseGrid,
    win: baseWin,
    multiplier: 1,
    scatterCount: baseScatters,
    triggeredFreeSpins: triggered,
    winDetails: baseDetails,
  });

  let totalWin = baseWin;

  // --- серия фриспинов ---
  let spinIndex = 1;
  let played = 0;

  while (remaining > 0 && played < MAX_FREE_SPINS) {
    remaining -= 1;
    played += 1;

    const stops: number[] = [];
    for (let reel = 0; reel < NUM_REELS; reel++) {
      stops.push(rng.nextInt(cfg.freeReels[reel].length));
    }
    const grid = windowFromStops(cfg.freeReels, stops);
    const evaluation = evaluateLines(cfg, grid);
    const scatters = countScatters(cfg, grid);
    const scatterMult = cfg.scatterPays[String(scatters)] ?? 0;
    const scatterPay = scatterMult * totalBet;

    const details: WinDetail[] = [...evaluation.details];
    if (scatterPay > 0) {
      details.push({
        symbol: cfg.scatter,
        count: scatters,
        pay: scatterMult * cfg.lines,
        scatter: true,
      });
    }

    let retriggered = 0;
    if (cfg.retriggerEnabled && scatters >= cfg.scatterTrigger) {
      // Ретриггер даёт ПОЛНУЮ награду спинов независимо от множителя.
      //
      // Так восстанавливается равенство матожиданий. Сжатая серия короче в
      // `bonusMultiplier` раз, значит и шансов поймать ретриггер у неё во
      // столько же раз меньше. Если делить награду на множитель, каждое
      // событие стоит столько же, сколько в обычной серии, а событий меньше —
      // и сжатые тиры недоплачивают. Замер на 2 млн раундов показывал
      // −0.34 п.п. Полная награда со множителем делает событие во столько же
      // раз дороже, во сколько оно реже.
      retriggered = cfg.freeSpinsAward[String(scatters)] ?? 0;
      remaining += retriggered;
    }

    // Множитель фриспинов умножает ЛИНИИ, но не выплату за scatter.
    // Так посчитан аналитический RTP; распространить множитель на scatter
    // значит незаметно раздать больше заявленного.
    // Множитель фриспинов умножает линии; множитель бонуса — весь спин
    // целиком, включая scatter. Иначе сжатая серия недоплачивала бы за
    // scatter ровно во столько раз, во сколько стала короче.
    const win = (evaluation.total * cfg.freeSpinMultiplier * betPerLine + scatterPay) * bonusMultiplier;

    spins.push({
      index: spinIndex,
      free: true,
      reelStops: stops,
      grid,
      win,
      multiplier: cfg.freeSpinMultiplier * bonusMultiplier,
      scatterCount: scatters,
      triggeredFreeSpins: retriggered,
      winDetails: details,
    });

    totalWin += win;
    spinIndex += 1;
  }

  // Потолок применяется к раунду целиком, а не к отдельному спину.
  const cap = cfg.maxWinCap * totalBet;
  const capped = totalWin > cap;
  if (capped) totalWin = cap;

  return {
    serverSeedHash: hashServerSeed(serverSeed),
    clientSeed,
    nonce,
    betPerLine,
    lines: cfg.lines,
    totalBet,
    totalWin,
    capped,
    drawCount: rng.drawCount,
    bonusTier,
    bonusMultiplier,
    spins,
  };
}

File: slotmath/round.py
"""
Эталонное проигрывание одного раунда — от сидов до итоговой выплаты.

Это ЕДИНСТВЕННОЕ место, где зафиксирован порядок обращений к RNG.
Всё остальное — сервер, клиент, офлайн-верификатор — обязано повторять
его байт в байт, иначе provably fair не работает.

Порядок жёстко определён так (см. docs/ROUND-PROTOCOL.md):

  1. Базовый спин: 5 обращений nextInt(len(baseReels[i])), i = 0..4.
  2. Если выпал триггер — фриспины разыгрываются ПО ОДНОМУ,
     каждый забирает 5 обращений nextInt(len(freeReels[i])), i = 0..4.
     Ретриггер увеличивает счётчик оставшихся спинов, порядок не меняет.

Модуль намеренно НЕ использует numpy: он должен читаться построчно и
переноситься на любой язык. Скорость здесь не важна — это эталон и
генератор фикстур, массовые прогоны делает `simulate.py`.

Совместимость с `src/engine/rng.ts` обеспечивается функциями
`hmac_block` и `RoundRandom` ниже: ключ HMAC — серверный сид как
UTF-8 строка (не как декодированные из hex байты), сообщение —
"clientSeed:nonce:cursor".
"""

from __future__ import annotations

import hashlib
import hmac
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

from .config import GameConfig
from .paylines import NUM_REELS, NUM_ROWS, PAYLINES

HMAC_BYTES = 32
BYTES_PER_FLOAT = 4
FLOATS_PER_BLOCK = HMAC_BYTES // BYTES_PER_FLOAT  # 8


def hash_server_seed(server_seed: str) -> str:
    """Коммитмент: SHA-256 от серверного сида, взятого как UTF-8 строка."""
    return hashlib.sha256(server_seed.encode("utf-8")).hexdigest()


def hmac_block(server_seed: str, client_seed: str, nonce: int, cursor: int) -> bytes:
    """Блок из 32 байт. Ключ — серверный сид как UTF-8, не как hex-байты."""
    message = f"{client_seed}:{nonce}:{cursor}".encode("utf-8")
    return hmac.new(server_seed.encode("utf-8"), message, hashlib.sha256).digest()


class RoundRandom:
    """Ленивый поток чисел одного раунда. Аналог класса из rng.ts."""

    def __init__(self, server_seed: str, client_seed: str, nonce: int):
        if ":" in client_seed:
            raise ValueError('clientSeed не может содержать ":"')
        self.server_seed = server_seed
        self.client_seed = client_seed
        self.nonce = nonce
        self.block_index = 0
        self.block = hmac_block(server_seed, client_seed, nonce, 0)
        self.offset = 0
        self.consumed = 0

    def next_float(self) -> float:
        if self.offset + BYTES_PER_FLOAT > HMAC_BYTES:
            self.block_index += 1
            self.block = hmac_block(
                self.server_seed, self.client_seed, self.nonce, self.block_index
            )
            self.offset = 0
        b = self.block
        o = self.offset
        self.offset += BYTES_PER_FLOAT
        self.consumed += 1
        return (
            b[o] / 256
            + b[o + 1] / 256 ** 2
            + b[o + 2] / 256 ** 3
            + b[o + 3] / 256 ** 4
        )

    def next_int(self, bound: int) -> int:
        if bound <= 0:
            raise ValueError(f"bound должен быть > 0, получено {bound}")
        return min(int(self.next_float() * bound), bound - 1)

    @property
    def draw_count(self) -> int:
        return self.consumed


# ---------------------------------------------------------------------------
# Оценка окна
# ---------------------------------------------------------------------------


def window_from_stops(reels: List[List[str]], stops: List[int]) -> List[List[str]]:
    """Окно 5x3 в виде window[reel][row]."""
    window: List[List[str]] = []
    for reel_index in range(NUM_REELS):
        strip = reels[reel_index]
        length = len(strip)
        stop = stops[reel_index]
        window.append([strip[(stop + row) % length] for row in range(NUM_ROWS)])
    return window


def evaluate_lines(
    cfg: GameConfig, window: List[List[str]]
) -> Tuple[int, List[dict]]:
    """
    Выплата по всем линиям в единицах ставки на линию + детализация.

    Правило классическое: слева направо от первого барабана, wild заменяет
    любой оплачиваемый символ, итог по линии — максимум по всем символам.
    """
    total = 0
    details: List[dict] = []

    for line_number, line in enumerate(PAYLINES, start=1):
        symbols = [window[reel][line[reel]] for reel in range(NUM_REELS)]

        best_pay = 0
        best_symbol: Optional[str] = None
        best_run = 0

        for symbol, tiers in cfg.paytable.items():
            run = 0
            for reel in range(NUM_REELS):
                if symbols[reel] == symbol or symbols[reel] == cfg.wild:
                    run += 1
                else:
                    break
            pay = tiers.get(run, 0) if run >= 3 else 0
            if pay > best_pay:
                best_pay = pay
                best_symbol = symbol
                best_run = run

        if best_pay > 0:
            total += best_pay
            details.append(
                {
                    "line": line_number,
                    "symbol": best_symbol,
                    "count": best_run,
                    "positions": [[reel, line[reel]] for reel in range(best_run)],
                    "pay": best_pay,
                }
            )

    return total, details


def count_scatters(cfg: GameConfig, window: List[List[str]]) -> int:
    return sum(1 for reel in window for symbol in reel if symbol == cfg.scatter)


# ---------------------------------------------------------------------------
# Раунд
# ---------------------------------------------------------------------------


@dataclass
class SpinRecord:
    index: int
    free: bool
    reel_stops: List[int]
    grid: List[List[str]]
    win: int
    multiplier: int
    scatter_count: int
    triggered_free_spins: int
    win_details: List[dict] = field(default_factory=list)

    def as_dict(self) -> dict:
        return {
            "index": self.index,
            "free": self.free,
            "reelStops": self.reel_stops,
            "grid": self.grid,
            "win": self.win,
            "multiplier": self.multiplier,
            "scatterCount": self.scatter_count,
            "triggeredFreeSpins": self.triggered_free_spins,
            "winDetails": self.win_details,
        }


@dataclass
class RoundRecord:
    server_seed_hash: str
    client_seed: str
    nonce: int
    bet_per_line: int
    lines: int
    total_bet: int
    total_win: int
    capped: bool
    draw_count: int
    spins: List[SpinRecord]

    def as_dict(self) -> dict:
        return {
            "serverSeedHash": self.server_seed_hash,
            "clientSeed": self.client_seed,
            "nonce": self.nonce,
            "betPerLine": self.bet_per_line,
            "lines": self.lines,
            "totalBet": self.total_bet,
            "totalWin": self.total_win,
            "capped": self.capped,
            "drawCount": self.draw_count,
            "spins": [s.as_dict() for s in self.spins],
        }


# Предохранитель от патологического зацикливания на ретриггерах.
# При принятой математике недостижим: вероятность ретриггера 0.38%,
# 200 фриспинов подряд имеют вероятность порядка 10^-500.
MAX_FREE_SPINS = 200


def play_round(
    cfg: GameConfig,
    server_seed: str,
    client_seed: str,
    nonce: int,
    bet_per_line: int = 1,
) -> RoundRecord:
    """Проигрывает раунд целиком: базовый спин плюс вся серия фриспинов."""
    rng = RoundRandom(server_seed, client_seed, nonce)
    total_bet = bet_per_line * cfg.lines
    spins: List[SpinRecord] = []

    # --- Базовый спин ---
    base_lengths = [len(r) for r in cfg.base_reels]
    stops = [rng.next_int(n) for n in base_lengths]
    window = window_from_stops(cfg.base_reels, stops)

    line_pay, details = evaluate_lines(cfg, window)
    scatter_count = count_scatters(cfg, window)
    scatter_pay = cfg.scatter_pays.get(scatter_count, 0) * total_bet

    if scatter_pay:
        details.append(
            {
                "symbol": cfg.scatter,
                "count": scatter_count,
                "pay": cfg.scatter_pays[scatter_count] * cfg.lines,
                "scatter": True,
            }
        )

    triggered = 0
    remaining = 0
    if scatter_count >= cfg.scatter_trigger:
        triggered = cfg.free_spins_award.get(scatter_count, 0)
        remaining = triggered

    base_win = line_pay * bet_per_line + scatter_pay
    spins.append(
        SpinRecord(
            index=0,
            free=False,
            reel_stops=stops,
            grid=window,
            win=base_win,
            multiplier=1,
            scatter_count=scatter_count,
            triggered_free_spins=triggered,
            win_details=details,
        )
    )
    total_win = base_win

    # --- Фриспины, строго по одному ---
    free_lengths = [len(r) for r in cfg.free_reels]
    spin_index = 1
    played = 0

    while remaining > 0 and played < MAX_FREE_SPINS:
        remaining -= 1
        played += 1

        f_stops = [rng.next_int(n) for n in free_lengths]
        f_window = window_from_stops(cfg.free_reels, f_stops)

        f_line_pay, f_details = evaluate_lines(cfg, f_window)
        f_scatter_count = count_scatters(cfg, f_window)
        f_scatter_pay = cfg.scatter_pays.get(f_scatter_count, 0) * total_bet

        if f_scatter_pay:
            f_details.append(
                {
                    "symbol": cfg.scatter,
                    "count": f_scatter_count,
                    "pay": cfg.scatter_pays[f_scatter_count] * cfg.lines,
                    "scatter": True,
                }
            )

        f_triggered = 0
        if cfg.retrigger_enabled and f_scatter_count >= cfg.scatter_trigger:
            f_triggered = cfg.free_spins_award.get(f_scatter_count, 0)
            remaining += f_triggered

        # Множитель применяется к линиям, но НЕ к прямой выплате за scatter.
        f_win = f_line_pay * cfg.free_spin_multiplier * bet_per_line + f_scatter_pay

        spins.append(
            SpinRecord(
                index=spin_index,
                free=True,
                reel_stops=f_stops,
                grid=f_window,
                win=f_win,
                multiplier=cfg.free_spin_multiplier,
                scatter_count=f_scatter_count,
                triggered_free_spins=f_triggered,
                win_details=f_details,
            )
        )
        total_win += f_win
        spin_index += 1

    # --- Потолок выигрыша ---
    cap = cfg.max_win_cap * total_bet
    capped = total_win > cap
    if capped:
        total_win = cap

    return RoundRecord(
        server_seed_hash=hash_server_seed(server_seed),
        client_seed=client_seed,
        nonce=nonce,
        bet_per_line=bet_per_line,
        lines=cfg.lines,
        total_bet=total_bet,
        total_win=total_win,
        capped=capped,
        draw_count=rng.draw_count,
        spins=spins,
    )
